### ISSUE TITLE
io.hitran.parse_global_quanta returning different datatypes depending on database formatting

### ISSUE THREAD

**Issue by dcmvdbekerom** [radis/radis#280]
### 🐛 Describe the bug

This error was found in the extension of fetching HITEMP database, detailed here: https://github.com/radis/radis/issues/210.
The CO2 HITEMP database comes in multiple separate files. 
When processing the 5th file (1000-1500cm-1), the following error occurs:

```
(114s)  30.1%   Parsed 3,374,322 / 11,193,608 lines. Wavenumber range 1000.00-1249.41 cm-1 is complete.Traceback (most recent call last):
  File "C:\Users\dcmvd\AppData\Local\Programs\Python\Python37\lib\site-packages\pandas\io\pytables.py", line 3990, in _get_blocks_and_items
    b, b_items = by_items.pop(items)
KeyError: ('A', 'El', 'Pshft', 'Tdpair', 'airbrd', 'gp', 'gpp', 'int', 'l2u', 'ru', 'selbrd', 'v1u', 'v2u', 'v3u')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test_spectrum_factory.py", line 11, in <module>
    sf.fetch_databank('hitemp',
  File "c:\users\dcmvd\documents\github\radis\radis\lbl\loader.py", line 975, in fetch_databank
    return_local_path=True,
  File "c:\users\dcmvd\documents\github\radis\radis\io\hitemp.py", line 373, in fetch_hitemp
    )
  File "C:\Users\dcmvd\AppData\Local\Programs\Python\Python37\lib\site-packages\pandas\io\pytables.py", line 1106, in put
    dropna=dropna,
  File "C:\Users\dcmvd\AppData\Local\Programs\Python\Python37\lib\site-packages\pandas\io\pytables.py", line 1755, in _write_to_group
    track_times=track_times,
  File "C:\Users\dcmvd\AppData\Local\Programs\Python\Python37\lib\site-packages\pandas\io\pytables.py", line 4224, in write
    data_columns=data_columns,
  File "C:\Users\dcmvd\AppData\Local\Programs\Python\Python37\lib\site-packages\pandas\io\pytables.py", line 3852, in _create_axes
    block_obj, table_exists, new_non_index_axes, self.values_axes, data_columns
  File "C:\Users\dcmvd\AppData\Local\Programs\Python\Python37\lib\site-packages\pandas\io\pytables.py", line 3998, in _get_blocks_and_items
    ) from err
ValueError: cannot match existing table structure for [A,El,Pshft,Tdpair,airbrd,gp,gpp,int,l2u,ru,selbrd,v1u,v2u,v3u] on appending data
```

### 🎯 Details
It is expected that adding the new dataframe to the existing database that is being built up will occur without issues. The fact that the first 4 files worked without issue appear to indicate it is related to the formatting of the HITEMP database file. We have seen formatting issues of HITEMP database before in https://github.com/radis/radis/issues/111, and the current issue is indeed related.

After downloading the database file, it is parsed by `io.hitran.parse_global_quanta` to get information on the vibrational quanta of the upper state. For some lines in the database, these quantum numbers are not defined and instead `io.hitran.parse_global_quanta` will assign NaN's. Because the return datatype of `io.hitran.parse_global_quanta` is not explicitly specified, it will return whatever it needs to accommodate the global quanta. Since these include NaN's every now and then, this will be `np.float64` most of the time. This is because integer types, which would be more suited for conveying quantum numbers, cannot store NaN's.

Now back to the file that is causing the error, it just so happens that in the last block of the 5th database file, there is _not a single line with missing upper level quanta_. This sounds like a good thing, but this means `io.hitran.parse_global_quanta` doesn't produce any NaN's, and the datatype returned is `np.uint32`. Since this is different from `np.float64`, the dataframe cannot be added without additional conversions, and an exception is thrown.

### 💡 Possible solutions
The biggest problem here is that the return datatype of `io.hitran.parse_global_quanta` isn't defined and can change depending on the database. The easy, inefficient solution is to just always return global quanta as `np.float64`. This is what I will implement in https://github.com/radis/radis/issues/210 for the time being.

It would be much more memory efficient to use integer types however. Since vibrational levels rarely go higher than 255, `np.uint8` seems suited. However, we still need to account for levels with absent upper level quantum numbers. That's why I think we should use `np.int8`, which is a singed type, that can use `-1` to indicate absent data (quantum numbers can normally never go below zero so there can be no confusion). `np.int8` has an upper limit of 127, which I think will be higher than any physical quantum number we would encounter. 

Note that we already wanted to explicitly type parts of the database in issue https://github.com/radis/radis/issues/118, so this adds some extra motivation. @gagan-aryan I think this would be a good issue to pick up for you as it's a relatively simple start into the efficient use of memory in the database!

**Comment by gagan-aryan** [comment]
@dcmvdbekerom this issue is currently blocked on #210 right? And to be clear, this is exclusive to just the HITEMP databases(CO2/H2O) with regard to converting it due to the presence of `NaN` values, but we need to the same for other species as well to reduce the memory usage. And the columns we need to convert to `np.int8` would be all the parameters with the type `integer` in this [pdf](https://hitran.org/hitemp/data/HITEMP-2010/CO2_line_list/Readme_for_CDSD.pdf)

**Comment by dcmvdbekerom** [comment]
@gagan-aryan I didn't get what you meant by :
> this issue is currently blocked on #210 right?
but other than that, you're right: this is (so far) exclusive to HITEMP, but it's possible other database formats we are not currently handling have the same issue. And indeed, other databases don't have this error, but they do need different types for memory efficiency. Indeed I would say all `integer` types in that manual can be `np.int8`, but `lower state j` should probably be `np.int16`. In fact, I don't think `lower state j` is ever absent in the database, so it would be safe to use `np.uint16`, but that's splitting hairs.

**Comment by dcmvdbekerom** [comment]
@erwanp I'm also having the same issue for H2O now. I noticed this issue came up in https://github.com/radis/radis/issues/211 and https://github.com/radis/radis/pull/264 before, but those fixes did not cover the current case.

I think we should think of a more comprehensive approach do dealing with these cases. I have three options we might consider:

1. If global/local quanta data is absent, throw an exception and add these lines to a new corresponding database that only stores equilibrium data. The name of the equilibrium database will be referenced to in the JSON entry for the current database. (so for example, we're fetching `HITEMP-CO2`, and it automatically makes a database `HITEMP-CO2-EQ-ONLY`.) When calculating equilibrium spectra, RADIS should see if such a database exists and load those lines as well.
2. The easy solution: just ignore the problematic lines and call it a day. Perhaps store those lines in a log somewhere and print a warning that these lines have been dropped.
3.  We do both 1 and 2 and let the user decide with a switch in the config (preferred option obviously)

It's possible that this issue is already addressed in some ways that I'm not aware of, in that case I didn't intend on reinventing the wheel.

**Comment by erwanp** [comment]
Do you already know why #264 didn't fix it ? Can you share a minimal code? 

By default, most RADIS user really only want equilibirum spectra ; so global/local quanta columns should be kept without being parsed, but the H2O database still loaded. This will ensure 95% of use-cases still work. 
We can add an informative parameter when registering the database in the radis.json  ("For equilibrium calculations only. Some lines are not identified ... ")

I do not like 2 ; dropping lines feels like we are editing the databases ; which should be really up-to the end user. At most, RADIS should point out the problematic lines

**Comment by dcmvdbekerom** [comment]
https://github.com/radis/radis/pull/264/files wraps `parse_local_quanta(df, mol)` in a `try/except` statement, but fundamentally `parse_local_quanta(df, mol)` still fails. It might be better to move the `try/except` to within `parse_local_quanta`.

**Comment by dcmvdbekerom** [comment]
>we should also never drop the global/local quanta column during the download of the database, even if it was parsed properly ; because some later spectral ranges may not be parsed as you explained in the issue ; in that case we'll need the full local/quanta column to be able to parse it at runtime.

Mmm I disagree here.. The errors occur when parsing empty lines, so not much use in keeping said empty lines..

**Comment by erwanp** [comment]
Right ! 
- if we are not able to parse the local quanta for a range of the database, we just retry and store this range without parsing local quanta.
- if the local quanta are parsed successfully, we also keep the local-quanta column in case other ranges of the database are not parsed.

**Comment by dcmvdbekerom** [comment]
It's a bit of a waste to stop parsing the entire range if there's only one problematic line..

I would suggest to do the parsing within a `try/except` statement, and if an exception occurs set the quanta of the failing lines to -1 and try to parse again. We should add a column to keep track whether the lines are parsed properly (`True/False`), and ignore the improperly parsed lines when plotting a non-eq spectrum.

**Comment by erwanp** [comment]
Setting to -1 is a great idea. 
Then only thing i want to avoid is to drop the line itself

**Comment by erwanp** [comment]
Being fixed in #306 , using the `-1` convention for unlabelled line. The following code works : 
```python
# CO2 over an overlapping range of missing lines / non missing lines
df = fetch_hitemp("CO2", load_wavenum_min=800, load_wavenum_max=1300, verbose=3,
                      cache='regen')
print(df.dtypes["v3u"])  # -> int64   Worked
```

Note: I found in the code comments I had written about how [H2O has some negative vibrational coefficients](https://github.com/radis/radis/blob/74c9883d0063060a3bca799a0753eb03b2312b3a/radis/io/hitran.py#L522) (?). To check when we want to implement nonequilibrium for H2O.
